From a782fda31d4524c3cf05cffee2123857b3e05ccd Mon Sep 17 00:00:00 2001
From: ctang <ctang@cloudera.com>
Date: Sun, 31 Jul 2016 18:11:04 -0400
Subject: [PATCH 0756/1164] CDH-44278: HIVE-14383: SparkClientImpl should pass
 principal and keytab to spark-submit instead of
 calling kinit explicitely (Chaoyu Tang, reviewed
 by Xuefu Zhang)

Change-Id: I940c90426b567c9d0c3e5b0e732b01a2f28ee378
---
 .../apache/hive/spark/client/SparkClientImpl.java  |   21 ++++++++++----------
 1 file changed, 10 insertions(+), 11 deletions(-)

diff --git a/spark-client/src/main/java/org/apache/hive/spark/client/SparkClientImpl.java b/spark-client/src/main/java/org/apache/hive/spark/client/SparkClientImpl.java
index 59a31c2..377b24d 100644
--- a/spark-client/src/main/java/org/apache/hive/spark/client/SparkClientImpl.java
+++ b/spark-client/src/main/java/org/apache/hive/spark/client/SparkClientImpl.java
@@ -328,17 +328,6 @@ public void run() {
 
       List<String> argv = Lists.newArrayList();
 
-      if ("kerberos".equalsIgnoreCase(hiveConf.get(HADOOP_SECURITY_AUTHENTICATION))) {
-          argv.add("kinit");
-          String principal = SecurityUtil.getServerPrincipal(hiveConf.getVar(ConfVars.HIVE_SERVER2_KERBEROS_PRINCIPAL),
-              "0.0.0.0");
-          String keyTabFile = hiveConf.getVar(ConfVars.HIVE_SERVER2_KERBEROS_KEYTAB);
-          argv.add(principal);
-          argv.add("-k");
-          argv.add("-t");
-          argv.add(keyTabFile + ";");
-      }
-
       if (sparkHome != null) {
         argv.add(new File(sparkHome, "bin/spark-submit").getAbsolutePath());
       } else {
@@ -376,6 +365,16 @@ public void run() {
         argv.add("org.apache.spark.deploy.SparkSubmit");
       }
 
+      if ("kerberos".equals(hiveConf.get(HADOOP_SECURITY_AUTHENTICATION))) {
+          String principal = SecurityUtil.getServerPrincipal(hiveConf.getVar(ConfVars.HIVE_SERVER2_KERBEROS_PRINCIPAL),
+              "0.0.0.0");
+          String keyTabFile = hiveConf.getVar(ConfVars.HIVE_SERVER2_KERBEROS_KEYTAB);
+          argv.add("--principal");
+          argv.add(principal);
+          argv.add("--keytab");
+          argv.add(keyTabFile);
+      }
+
       if (master.equals("yarn-cluster")) {
         String executorCores = conf.get("spark.executor.cores");
         if (executorCores != null) {
-- 
1.7.9.5

